rfc-120-csv-preview.md

```diff
@@ -0,0 +1,55 @@
+# CSV Preview
+
+## Summary
+
+rove the 'CSV Preview' feature out of Whitehall and into Government Frontend, changing the structure of the associated URLs so they can also be used with documents published by Content Publisher.
+
+
+## Problem
+
+A user viewing a Whitehall document on GOV.UK can click a "View Online" link for any of its CSV file attachments. This takes the user to a new page, currently rendered by Whitehall. The page contains:
+
+   - A 'header' section with details of the parent document (and navigation links)
+   - A 'body' section with a preview of the content of the CSV attachment
+
+This 'CSV Preview' feature only works for CSV file attachments uploaded through Whitehall. We need it to work for CSV attachments published through Content Publisher, which is superseding Whitehall.
+
+> We do not expect this feature to be used for other (non-CSV) attachments, for other document types or by other publishing apps. If this need arises in future, we should review the decision of this RFC.
+
+## Proposal
+
+Re-implement the 'CSV Preview' feature in Government Frontend. Change existing attachments to use Government Frontend for CSV Preview, and deprecate/delete the feature in Whitehall.
+
+Government Frontend currently renders all documents published by Whitehall, that have CSV file attachments. This makes it a relevant codebase to implement this feature.
+
+We expect the implementation to be small, isolated and not to grow/change much. Adding a new responsibility to Government Frontend seems like a reasonable compromise when compared to the [rejected proposals](#rejected-proposals).
+
+### URL Structure
+
+We will need to use a different URL structure, since the current one is coupled to Whitehall. The URL needs to include information about the parent document and the attachment to preview.
+
+```
+# current URL example
+https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/856932/Management_information_-_further_education_and_skills_-_as_at_31_December_2019.csv/preview
+
+# proposed URL example
+https://www.gov.uk/government/remaining-parent-document-base-path/csv-preview/asset-id
+```
+
+> Note: the `asset-id` should match the `id` of an asset in the `attachments` array, as defined by [RFC 116][rfc-116]. The `attachments` array contains the URL needed to download the asset in order to preview it.
```

Comment:
Currently the `id` is used (or is planned to be used) for the rendering app to fetch the metadata of the featured attachments, which means changing the `id` in a new edition is fine and doesn't break anything.  This is also kind of the case for content IDs: in principle we could change the content ID of a document and, as long as we updated the `links` of everything which referred to it, everything would just work.

This proposed change means we can no longer change the `id`, as public URLs will stop working.  This is fine, but something to be aware of going forwards, as I could see it potentially catching people out.

Reply:
Thanks for comments @barrucadu - this is a good point and something I hadn't thought of. After comparing two potential approaches, I can see arguments both ways:

   - **Using the `asset-id`** (as [badly defined in Asset Manager](https://github.com/alphagov/asset-manager/tree/7cd3d543e3b44ba3bcb276d8c9aef11eadd3bb1a#create-an-asset) - it's just the string on the end). This would mean we match the behaviour of the asset (replace/redirect), at the expense of an extra API query to get the asset URL.

   - **Using the `attachment-id`** (as per RFC 116). This saves an API query, since the payload contains the asset URL. Changing the ID (e.g. by replacing the file) would cause any external links to 404, unlike the `asset-id` approach.

Since we're not using an `assets` URL for CSV Preview, it seems OK to have a different behaviour. In both cases, a deleted asset would cause a 404. Perhaps a general approach to address this (hopefully infrequent) concern is to unblock the user by showing a more helpful error page, with a link (or just an automatic redirect) to the parent document.

I'll make a note of this to consider when we re-implement. 

Do you think it should be included in the RFC?

(One more minor advantage of the `attachment-id` approach is that it's more likely to be human-readable, at least for attachments published by Content Publisher.)

Reply:
:+1:

I don't think this needs adding to the RFC, as I don't think rendering in Government Frontend but still querying Asset Manager is a good idea, it's almost the worst of both worlds!